### PR TITLE
randomness #8: framework update from randomnet

### PR DESCRIPTION
**Pull request by zjma**

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/configs/consensus_config.move [hunk 1]
```diff
@@ -23,7 +25,11 @@ module aptos_framework::consensus_config {
         move_to(aptos_framework, ConsensusConfig { config });
     }
 
-    /// This can be called by on-chain governance to update on-chain consensus configs.
+    /// Deprecated by `set_for_next_epoch()`.
+    ///
+    /// WARNING: calling this while randomness is enabled will trigger a new epoch without randomness!
```
Comment (danielxiangzl) [inline aptos-move/framework/aptos-framework/sources/configs/consensus_config.move:30]: Should we add checks in `set` and `set_for_next_epoch`, to make sure the conditions are met? like `set` should be called only when randomness is not enabled. If a governance proposal mistakenly uses `set`, this prevents entering a new epoch without randomness. On the other hand, using `set_for_next_epoch` without randomness enabled will never change epoch?

File: aptos-move/framework/aptos-framework/sources/jwks.move [hunk 2]
```diff
@@ -184,20 +186,70 @@ module aptos_framework::jwks {
...
+    /// Example usage:
+    /// ```
+    /// aptos_framework::jwks::upsert_oidc_provider_for_next_epoch(
+    ///     &framework_signer,
+    ///     b"https://accounts.google.com",
+    ///     b"https://accounts.google.com/.well-known/openid-configuration"
+    /// );
+    /// aptos_framework::aptos_governance::reconfigure(&framework_signer);
+    /// ```
+    public fun upsert_oidc_provider_for_next_epoch(fx: &signer, name: vector<u8>, config_url: vector<u8>): Option<vector<u8>> acquires SupportedOIDCProviders {
```
Comment (danielxiangzl) [inline aptos-move/framework/aptos-framework/sources/jwks.move:199]: Does upserting or deleting oidc trigger epoch change?

File: aptos-move/framework/aptos-framework/sources/reconfiguration_state.move [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            })
+        }
+    }
+
+    public fun initialize_for_testing(fx: &signer) {
+        initialize(fx)
+    }
+
+    /// Return whether the reconfiguration state is marked "in progress".
+    public(friend) fun is_in_progress(): bool acquires State {
```
Comment (danielxiangzl) [inline aptos-move/framework/aptos-framework/sources/reconfiguration_state.move:53]: Where is this used?

File: aptos-move/framework/aptos-framework/sources/reconfiguration_state.move [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    use aptos_framework::timestamp;
+
+    friend aptos_framework::reconfiguration;
+    friend aptos_framework::reconfiguration_with_dkg;
+    friend aptos_framework::stake;
+
+    const ERECONFIG_NOT_IN_PROGRESS: u64 = 1;
+
+    /// Reconfiguration drivers update this resources to notify other modules of some reconfiguration state.
+    struct State has key {
```
Comment (danielxiangzl) [inline aptos-move/framework/aptos-framework/sources/reconfiguration_state.move:19]: To double check, this is introduced for locking the validator set in stake.move?

File: aptos-move/framework/aptos-framework/sources/jwks.move (hunk 2 above)
Comment (zjma) [reply aptos-move/framework/aptos-framework/sources/jwks.move:199]: i think it should

File: aptos-move/framework/aptos-framework/sources/configs/consensus_config.move (hunk 1 above)
Comment (zjma) [reply aptos-move/framework/aptos-framework/sources/configs/consensus_config.move:30]: `set_for_next_epoch(); reconfigure();` combination is supposed to be the only API.
. but we have some tests relying on `set()`, as well as some genesis usage, so currently it's hard to even update them.

But yes, eventually it should be what you described.

File: aptos-move/framework/move-stdlib/sources/configs/features.move [hunk 5]
```diff
@@ -392,6 +392,15 @@ module std::features {
         is_enabled(CONCURRENT_FUNGIBLE_ASSETS) && aggregator_v2_api_enabled()
     }
 
+    /// The on-chain randomness feature.
+    ///
+    /// Lifetime: transient
+    const RECONFIGURE_WITH_DKG: u64 = 45;
```
Comment (danielxiangzl) [inline aptos-move/framework/move-stdlib/sources/configs/features.move:398]: hmmm both `JWK_CONSENSUS` and `CONCURRENT_FUNGIBLE_ASSETS` is 49..

File: aptos-move/framework/aptos-framework/sources/validator_consensus_info.move [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+module aptos_framework::validator_consensus_info {
+    /// Information about a validator that participates consensus.
+    struct ValidatorConsensusInfo has copy, drop, store {
+        addr: address,
+        pk_bytes: vector<u8>,
+        voting_power: u64,
+    }
+
+    /// Create a default `ValidatorConsensusInfo` object. Value may be invalid. Only for place holding prupose.
+    public fun default(): ValidatorConsensusInfo {
```
Comment (danielxiangzl) [inline aptos-move/framework/aptos-framework/sources/validator_consensus_info.move:11]: where is `default` used?

File: aptos-move/framework/src/natives/randomness.rs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+use better_any::{Tid, TidAble};
+use move_vm_runtime::native_functions::NativeFunction;
+use move_vm_types::{loaded_data::runtime_types::Type, values::Value};
+use smallvec::{smallvec, SmallVec};
+use std::collections::VecDeque;
+
+/// A txn-local counter that increments each time a random 32-byte blob is requested.
+#[derive(Tid, Default)]
+pub struct RandomnessContext {
+    txn_local_state: Vec<u8>, //   8-byte counter
```
Comment (zekun000) [inline aptos-move/framework/src/natives/randomness.rs:15]: do we need to include some txn specific information like txn hash?

File: aptos-move/framework/src/natives/randomness.rs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+impl RandomnessContext {
+    pub fn new() -> Self {
+        Self {
+            txn_local_state: vec![0; 8],
+        }
+    }
+
+    pub fn increment(&mut self) {
+        for byte in self.txn_local_state.iter_mut() {
```
Comment (zekun000) [inline aptos-move/framework/src/natives/randomness.rs:26]: this is awkward lol, why not just use a u64? and to_le/be_bytes when have to return?

File: aptos-move/framework/aptos-framework/sources/stake.move [hunk 9]
```diff
@@ -431,29 +434,38 @@ module aptos_framework::stake {
...
     ) acquires ValidatorSet {
+        assert_reconfig_not_in_progress();
         system_addresses::assert_aptos_framework(aptos_framework);
-
         let validator_set = borrow_global_mut<ValidatorSet>(@aptos_framework);
         let active_validators = &mut validator_set.active_validators;
         let pending_inactive = &mut validator_set.pending_inactive;
-        let len = vector::length(validators);
+        spec {
+            update ghost_active_num = len(active_validators);
```
Comment (zekun000) [inline aptos-move/framework/aptos-framework/sources/stake.move:443]: what does ghost mean?

File: aptos-move/framework/aptos-framework/sources/stake.spec.move [hunk 10]
```diff
@@ -602,10 +656,12 @@ spec aptos_framework::stake {
     spec add_stake_with_cap {
         include ResourceRequirement;
         let amount = coins.value;
+        aborts_if reconfiguration_state::spec_is_in_progress();
         include AddStakeWithCapAbortsIfAndEnsures { amount };
     }
 
     spec add_stake {
+        aborts_if reconfiguration_state::spec_is_in_progress();
```
Comment (zekun000) [inline aptos-move/framework/aptos-framework/sources/stake.spec.move:664]: so we still don't have apply to all except in the spec?

Comment (zjma) [reply aptos-move/framework/aptos-framework/sources/stake.spec.move:664]: let's see what @rahxephon89  says

File: aptos-move/framework/aptos-framework/sources/stake.move (hunk 9 above)
Comment (zjma) [reply aptos-move/framework/aptos-framework/sources/stake.move:443]: let's see what @rahxephon89  says

File: aptos-move/framework/aptos-framework/sources/stake.spec.move (hunk 10 above)
Comment (rahxephon89) [reply aptos-move/framework/aptos-framework/sources/stake.spec.move:664]: Will revisit all related cases after it is merged to main

File: aptos-move/framework/aptos-framework/sources/stake.move (hunk 9 above)
Comment (rahxephon89) [reply aptos-move/framework/aptos-framework/sources/stake.move:443]: This is used to prove that len(active_validators) + len(pending_inactive) is a invariant (does not change) during execution of remove_validators

File: aptos-move/framework/src/natives/randomness.rs (hunk 7 above)
Comment (zjma) [reply aptos-move/framework/src/natives/randomness.rs:15]: txn hash is currently involved [here](https://github.com/aptos-labs/aptos-core/blob/9e7cf6345a3ecc02576aeffea8f3d8a6b7ce1eee/aptos-move/framework/aptos-framework/sources/randomness.move#L59).

should be enough?

**Review by zekun000 — APPROVED** [review]

**Review by danielxiangzl — APPROVED** [review]

**Pull request merged**